[docs] document logfmt pipeline stage (#6238)
grafana/loki: docs/sources/clients/promtail/stages/logfmt.md

---

> title: logfmt

Reviewer: Do you mind adding `logfmt` to the list of existing stages in the stages page? (this one: https://grafana.com/docs/loki/latest/clients/promtail/pipelines/)

> title: logfmt
> ---
> # `logfmt` stage

Reviewer 2: As we add new content to the Loki documentation, we're trying to standardize what we include in the YAML front matter.  This suggestion adds 2 new fields to implement that.  I'm leaving out the weight field, as adding it only to this stage's file would mess up our (fine) alphabetized organization.

```suggestion
title: logfmt
menuTitle: logfmt
description: The logfmt parsing stage reads logfmt log lines and extracts the data into labels.
---
# `logfmt` stage
```

The `logfmt` stage is a parsing stage that reads the log line as [logfmt](https://brandur.org/logfmt) and allows extraction of data into labels.

## Schema

```yaml
logfmt:

>   # Set of key/value pairs for mapping of logfmt fieds to extracted labels. The key will be
>   # the key in the extracted data while the expression will be the value. If the value
>   # is empty then the logfmt field with the same name is extracted.

Reviewer 2: What do you think of being more explicit?
```suggestion
  # Set of key/value pairs for mapping of logfmt fields to extracted labels. The YAML key will be
  # the key in the extracted data, while the expression will be the YAML value. If the value
  # is empty, then the logfmt field with the same name is extracted.
```

  mapping:
    [ <string>: <string> ... ]

  # Name from extracted data to parse. If empty, uses the log message.
  [source: <string>]
```

> This stage uses the [go-logfmt](https://github.com/go-logfmt/logfmt) unmarshaler, which means non-string types like
> numbers or booleans will be unmarshaled into those types. The extracted data
> can hold non-string values and this stage does not do any type conversions;
> downstream stages will need to perform correct type conversion of these values
> as necessary. Please refer to the [the `template` stage](../template/) for how
> to do this.
> 
> If the value extracted is a complex type it's value is extracted as a string.

Reviewer 2: Just a bit of grammar and spelling suggestions.

```suggestion
This stage uses the [go-logfmt](https://github.com/go-logfmt/logfmt) unmarshaler, which means non-string types like
numbers or booleans will be unmarshaled into those types. The extracted data
can hold non-string values, and this stage does not do any type conversions;
downstream stages will need to perform correct type conversion of these values
as necessary. Please refer to the [`template` stage](../template/) for how
to do this.

If the value extracted is a complex type, its value is extracted as a string.
```

## Examples

### Using log line

For the given pipeline:

```yaml

> - logfmt:

Reviewer: Do you mind wrapping this logfmt inside a `pipeline_stages` key?

Reviewer: agreed.

    mapping:
      timestamp: time
      app:
      duration:

>       unknown:

Reviewer: Do you mind using the `labels` stage in conjunction and adding a disclaimer telling that if not listed under `labels`, extracted labels are to be ignored?

```

Given the following log line:

```
time=2012-11-01T22:08:41+00:00 app=loki level=WARN duration=125 message="this is a log line" extra="user=foo""
```

The following key-value pairs would be created in the set of extracted data:

- `timestamp`: `2012-11-01T22:08:41+00:00`
- `app`: `loki`
- `duration`: `125`

### Using extracted data

For the given pipeline:

```yaml
- logfmt:
    mapping:
      extra:
- logfmt:
    mapping:
      user:
    source: extra
```

And the given log line:

```

> time=2012-11-01T22:08:41+00:00 app=loki level=WARN duration=125 message="this is a log line" extra="user=foo""

Reviewer 2: Is there an extra double quote mark at the end of this log line?  If yes, this suggestion removes it.
```suggestion
time=2012-11-01T22:08:41+00:00 app=loki level=WARN duration=125 message="this is a log line" extra="user=foo"
```
